### ISSUE TITLE
torch.scatter_: support index.size(d) > src.size(d)

### ISSUE THREAD

**Issue by ekhahniii** [pytorch/pytorch#63265]
## 🚀 Feature
Teach [torch.Tensor.scatter_](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html) to handle `index.size(d) > src.size(d)`.

## Motivation

Currently, `torch.Tensor.scatter_` requires `index.size(d) <= src.size(d)` for all dimensions `d`, _unless_ `src` is float-valued. This constraint seems artificial.

```
import math
import torch

device = 'cuda'  # fails if device == 'cpu' (cf. #61854)

dim = 1
shape_src = 3, 5
shape_idx = 3, 7

index = torch.empty(shape_idx, dtype=torch.long, device=device).random_(0, shape_src[dim])

# succeeds
output = torch.zeros(shape_src, device=device)
output.scatter_(dim, index, 1.0, reduce='add')
print(output)

# desired usage
# output = torch.zeros(shape_src, device=device)
# values = torch.rand(shape_src, device=device)
# output.scatter_(dim, index, values, reduce='add')   # <-- failure
# print(output)
```

## Alternatives

An inefficient workaround is to chunk the index tensor:
```
output = torch.zeros(shape_src, device=device)
values = torch.ones(shape_src, device=device)
chunks = math.ceil(index.shape[dim] / output.shape[dim])
for index_chunk in torch.chunk(index, chunks, dim):
    output.scatter_(dim, index_chunk, values, reduce='add')
print(output)
```

**Comment by bfgray3** [comment]
I'd like to work on this issue.

**Comment by dhivyadharshin** [comment]
Can I solve this issue?

**Comment by adrian2504** [comment]
Would like to work on this

**Comment by MohitAnand-Hub** [comment]
You are looking to index on different dimensions at the same time. I had a look around in the documentation, torch.index_add will only receive a vector as index. My hopes were on torch.scatter but it doesn't to fit well to this problem. As it turns out you can achieve this pretty easily with a little work, the most difficult parts are the setup and teardown.

**Comment by dhivyadharshin** [comment]
Awesome, thanks!

**Comment by NaYrA-IaR** [comment]
Hello, I am a beginner and I would like to contribute to this project. Can you please help me with how to get started ?

**Comment by goelm08** [comment]
Hello, I am a beginner and I want to contribute to this project. Can you please help me to get started?

**Comment by ankit332000** [comment]
Hello , I wanna do my first pull request can you help me with that by allowing me to contribute?

**Comment by dhivyadharshin** [comment]
Yes, I can help you.

**Comment by Kamalesh3112** [comment]
I would like to solve this issue , can I anyone help me solve this issue?

**Comment by ankit332000** [comment]
Hi,
How to start can you help me out?

**Comment by muskan117** [comment]
i would like to work on this

**Comment by saireddythfc** [comment]
I would like to resolve this issue. Can you help me start with this?

**Comment by Prashant723** [comment]
I would like to contribute to this as my first open source contribution. Can you help me out to start with this?

**Comment by Ayush331** [comment]
I am new to opensource ,I would like to resolve this and contribute to this ?Can I

**Comment by hardyjoshi17** [comment]
Hello, I am new to opensource and want to contribute to opensource. Can someone help me with it?

**Comment by dhivyasreedhar** [comment]
Hi, can I work on this? I'm new so, can someone guide me?

**Comment by Aravind-11** [comment]
Hello. I'm learning open source, would be great if someone could elaborate the issue.

**Comment by vikramkanishka** [comment]
> 
> 
> I'd like to work on this issue.

can anyone help me to solve this issue??

**Comment by bfgray3** [comment]
i had started #63452 but i won't be able to complete it.

**Comment by Sirish07** [comment]
Hey everyone,

I noticed that the previous PR #63452 is closed. I am working on this issue.

**Comment by viraatdas** [comment]
Hey @Sirish07 wondering what your progress on this is? Interested in seeing if I can provide any assistance to get this done as well.

**Comment by redmagnos** [comment]
@viraatdas Don't know if you're still interested, but I'm giving it a go here and while it seems like index.size(dim) > src.size(dim) is pretty straightforward, other values of d are giving me more of a headache.

Also, the "inefficient workaround" in the problem statement is similarly restricted to only working along a single specific d (there specified to be = dim). You'd need further for loops if index.size(d) > src.size(d) for more dimensions than just the one, if I understand this correctly.

**Comment by redmagnos** [comment]
Tried to do something but having trouble with calling arrays from within the CUDA loop, probably something obvious that I'm missing.

**Comment by redmagnos** [comment]
Believe I've fixed it.

**Comment by keshav0306** [comment]
hey i want to tackle this issue, but can only do so and test for device='cpu' and will be working on only the cpu kernel. is it fine?

**Comment by samhu1** [comment]
Is this issue still open?
